app/controllers/experiments_controller.rb

```diff
@@ -410,50 +419,31 @@ def create
     # Params validation
     type = params[:type] || 'experiment'
     Utils::raise_error_unless_has_key(CONSTRUCTORS, type, "Not a correct experiment type: #{type}")
-    workers_scaling_params = nil
+
     workers_scaling_enabled = (params[:workers_scaling] == 'true')
+    workers_scaling_initializer = nil
     if workers_scaling_enabled
-      message_prefix = 'Missing workers scaling parameter'
-      Utils::raise_error_unless_has_key(params, :workers_scaling_params, message_prefix.pluralize)
-      workers_scaling_params = Utils::parse_json_if_string(params[:workers_scaling_params]).symbolize_keys
-      if workers_scaling_params[:plgrid_default]
-        if InfrastructureFacadeFactory.get_facade_for(:qsub).get_infrastructure_configurations(current_user.id).blank?
-          raise InfrastructureErrors::NoCredentialsError.new('Missing credentials for PlGrid resources')
-        end
-      else
-        [:name, :allowed_infrastructures, :time_limit].each do |param|
-          Utils::raise_error_unless_has_key(workers_scaling_params, param, "#{message_prefix} #{param}",
-                                            'workers_scaling_params')
-        end
-      end
-      # TODO more precise validation
+      workers_scaling_initializer = WorkersScaling::Initializer.new(current_user.id, params)
+      workers_scaling_initializer.validate_params
     end
 
     # Create experiment
-    experiment = send(CONSTRUCTORS[type])
+    result = send(CONSTRUCTORS[type])
 
     # Start workers scaling
-    if workers_scaling_enabled
-      planned_finish_time = Time.now + (workers_scaling_params[:time_limit] || 0).minutes
-      if workers_scaling_params[:plgrid_default]
-        WorkersScaling::AlgorithmFactory.plgrid_default(experiment.id.to_s, current_user.id)
-        experiment.plgrid_default = true
-      else
-        allowed_infrastructures = workers_scaling_params[:allowed_infrastructures].map do |record|
-          # TODO Introduce infrastructure id
-          {infrastructure: {name: record['name'].to_sym, params: record['params'].symbolize_keys}, limit: record['limit']}
-        end
-        WorkersScaling::AlgorithmFactory.initial_deployment(
-            experiment.id,
-            current_user.id,
-            allowed_infrastructures,
-            planned_finish_time,
-            workers_scaling_params
-        )
-      end
-      experiment.workers_scaling = true
-      experiment.planned_finish_time = planned_finish_time
-      experiment.save
+    if workers_scaling_enabled and result[:status] == :ok
+      workers_scaling_initializer.start(result[:experiment])
+    end
+
+    if not result.has_key? :html and not result.has_key? :json
+      flash[:error] = 'Unexpected error, please contact administrator'
+      result[:html] = experiments_path
+      result[:json] = {status: :error, message: flash[:error]}
+    end
+
+    respond_to do |format|
+      format.html { redirect_to(result[:html]) } if result.has_key? :html
```

Comment:
Odpowiem tutaj zbiorczo na 3 powyższe pytania bo dotyczą tej samej kwestii.

> Czy nie mozna by bylo tutaj wykorzystac experiment_factory ? Tablica konstruktorow wyglada jak fabryka.

To co tutaj się dzieje jest rozwiązaniem problemu który kiedyś istniał, to znaczy tworzenia eksperymentu przez 3 różne zapytania (osobne na zwykły, custom point i supervised). Zostało to zunifikowane pod jedno zapytanie które po typie eksperymentu wywołuje tworzenie odpowiedniego z nich. Tworzenie różnych typów eksperymentów zawiera więcej logiki niż samo wywołanie fabryki. Modyfikacja tego podejścia nie jest związana ze skalowaniem Workerów.

> Dlaczego nie mozna tutaj dac "experiments_path" ? I dlaczego caly format.html jest za ifem ?
> Analogiczne pytanie dotyczy jsona ponizej.

To co tutaj się dzieje jest rozwiązaniem problemu pojawiającego się przy zastosowaniu podejścia opisanego powyżej. Ze względu na zaszłości historyczne każda z trzech metod tworzenia eksperymentu sama wywoływała respond_to. Jest to wysoce problematyczne gdy jest potrzeba wykonania kodu po utworzeniu eksperymentu, w tym przypadku wystartowania skalowania Workerów. Może ono potencjalnie spowodować wyjątek, co przy poprzedniej implementacji prowadziło do ponownego wywołania respond_to. Dodatkowo nie powinno się startować skalowania gdy nie uda się utworzyć eksperymentu. Zwróciłem uwagę na ten problem  w pokoju WSS na hipchatcie. Rozwiązanie zaproponowane przez Jakuba: 

> każda metoda create, zamiast sama wywoływać respond_to, zwraca tego hasha jako return - jak mamy status != 'ok' to wtedy nie powinno wywoływać się uruchamiania algorytmu skalowania

Czyli, zamiast wywoływać samemu respond_to, metoda zwraca to co chciałaby żeby się tam znalazło, a samo wywołanie jest przeniesione na koniec metody create. Przed wywołaniem respond_to można teraz dodać startowanie skalowania Workerów bez obaw o błędy. 
> Dlaczego nie mozna tutaj dac "experiments_path" ?

Ponieważ nie w każdym przypadku metody przekierowują do experiments_path
> I dlaczego caly format.html jest za ifem ?

Ponieważ nie każda metoda wywoływała respond_to dla format.html, np create_custom_points_experiment
> Analogiczne pytanie dotyczy jsona ponizej.

W tym przypadku każda z metod wywoływała respond_to  dla format.json, jednak dla bezpieczeństwa jest to sprawdzane, potencjalnie metoda może nie zwracać odpowiedzi dla tego formatu. 

> Odpowiedz po wyjatku wyglada na taka sama jak w przypadku warunku unless flash[:error].blank? wyzej. Czy nie mozna tego jakos polaczyc aby nie duplikowac kodu ?

Odpowiedź rzeczywiście wygląda na taką samą i wskazane byłoby połączenie, jednak w tym zadaniu chodziło o przeniesienie wywołania respond_to do create, a nie zmiany wewnętrznej logiki każdej z metod tworzących eksperyment.